### PR TITLE
Install Qrack from source on build

### PR DESCRIPTION
**Pull request by WrathfulSpatula**
In theory, Qrack has a perfectly "sane and rational" set of various supported methods by which users can install it; in practice, this a large cognitive load on users who just want the Python package from PyPi. Hence, this PR adds an automatic Qrack source checkout and build for the Catalyst device, such that no manual installation of Qrack by the user is necessary to install the Catalyst Qrack device package from PyPi.

### CODE REVIEW

**Comment by cosenal** [comment]
I tried running `pip install -e .` on this branch and I get the following error:

```
-- Found OpenCL: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/System/Library/Frameworks/OpenCL.framework (found version "1.2")
    -- Configuring done (2.6s)
    -- Generating done (0.0s)
    -- Build files have been written to: /Users/cosenal/Coding/unitaryfund/pennylane-qrack/_skbuild/macosx-14.0-arm64-3.11/cmake-build
    [ 50%] Building CXX object CMakeFiles/qrack_device.dir/pennylane_qrack/qrack_device.cpp.o
    /Users/cosenal/Coding/unitaryfund/pennylane-qrack/pennylane_qrack/qrack_device.cpp:4:10: fatal error: 'QuantumDevice.hpp' file not found
    #include <QuantumDevice.hpp>
             ^~~~~~~~~~~~~~~~~~~
    1 error generated.
    make[2]: *** [CMakeFiles/qrack_device.dir/pennylane_qrack/qrack_device.cpp.o] Error 1
    make[1]: *** [CMakeFiles/qrack_device.dir/all] Error 2
    make: *** [all] Error 2
    Traceback (most recent call last):
      File "/opt/miniconda3/envs/pennylane-qrack/lib/python3.11/site-packages/skbuild/setuptools_wrap.py", line 674, in setup
        cmkr.make(make_args, install_target=cmake_install_target, env=env)
      File "/opt/miniconda3/envs/pennylane-qrack/lib/python3.11/site-packages/skbuild/cmaker.py", line 697, in make
        self.make_impl(clargs=clargs, config=config, source_dir=source_dir, install_target=install_target, env=env)
      File "/opt/miniconda3/envs/pennylane-qrack/lib/python3.11/site-packages/skbuild/cmaker.py", line 742, in make_impl
        raise SKBuildError(msg)

    An error occurred while building with CMake.
      Command:
        /usr/local/bin/cmake --build . --target install --config Release --
      Install target:
        install
      Source directory:
        /Users/cosenal/Coding/unitaryfund/pennylane-qrack
      Working directory:
        /Users/cosenal/Coding/unitaryfund/pennylane-qrack/_skbuild/macosx-14.0-arm64-3.11/cmake-build
    Please check the install target is valid and see CMake's output for more information.

    [end of output]

note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by WrathfulSpatula** [comment]
> I tried running pip install -e . on this branch and I got the following error:...

Likely, the issue is that, before packaging, you need to run these two commands, first, to pull in Qrack and Catalyst:
```
git submodule init
git submodule update
```

When the source package builds, the necessary files will end up copied over directly into the source package, so PyPi install doesn't need the submodules. Just a "packaging environment" does.

**Comment by WrathfulSpatula** [comment]
Sorry for the  "noise" in the diff from running `black`, but it shouldn't happen when we run it again.

**Comment by cosenal** [comment]
> > I tried running pip install -e . on this branch and I got the following error:...
> 
> Likely, the issue is that, before packaging, you need to run these two commands, first, to pull in Qrack and Catalyst:
> 
> ```
> git submodule init
> git submodule update
> ```
> 
> When the source package builds, the necessary files will end up copied over directly into the source package, so PyPi install doesn't need the submodules. Just a "packaging environment" does.

For that, don't we need to add qrack url in https://github.com/unitaryfund/pennylane-qrack/blob/qrack_source_install/.gitmodules ?

File: setup.py [hunk 1]
```diff
@@ -12,19 +12,25 @@
...
 
 
 with open("./pennylane_qrack/_version.py") as f:
     (version,) = re.findall('__version__ = "(.*)"', f.read())
 
+if not os.path.exists("qrack/_build"):
+    os.mkdir("qrack/_build")
+os.chdir("qrack/_build")
+os.system("cmake -DCPP_STD=14 .. ; make all -j(nproc --all)")
+os.chdir("../..")
```
Comment (vprusso) [inline setup.py:27]: It is usually recommended (for OS compatibility to do `os.chdir(os.path.join("..", ".."))` over `os.chdir("../..")`.

File: pennylane_qrack/qrack_device.py [hunk 2]
```diff
@@ -195,7 +208,9 @@ def apply_operations(self, operations):
             elif isinstance(op, QubitUnitary):
                 if len(op.wires) > 1:
                     raise DeviceError(
-                        "Operation {} is not supported on a {} device, except for single wires.".format(op.name, self.short_name)
+                        "Operation {} is not supported on a {} device, except for single wires.".format(
```
Comment (vprusso) [inline pennylane_qrack/qrack_device.py:211]: I find "f-strings" to be a bit more Pythonic:

```py
raise DeviceError(f"Operation {op.name} is not supported on a {self.short_name} device, except for single wires.")
```

**Review by vprusso — APPROVED** [review]
(minor changes, but aside from the LGTM!)

File: pennylane_qrack/qrack_device.py [hunk 3]
```diff
@@ -272,32 +287,71 @@ def _apply_gate(self, op):
...
 
         # translate op wire labels to consecutive wire labels used by the device
-        device_wires = self.map_wires((op.control_wires + op.wires) if op.control_wires else op.wires)
+        device_wires = self.map_wires(
+            (op.control_wires + op.wires) if op.control_wires else op.wires
+        )
         par = op.parameters
 
-        if opname in [''.join(p) for p in it.product(["Toffoli", "C(Toffoli)", "CNOT", "C(CNOT)", "MultiControlledX", "C(PauliX)"], ["", ".inv"])]:
+        if opname in [
```
[the remark this answers is not in the document]
Comment (WrathfulSpatula) [reply pennylane_qrack/qrack_device.py:299]: Thanks! I appreciate the suggestion, but, for my own sense of code style, if it isn't too long to be unreadable, and if it isn't at all reused, I'd rather not have it as a variable in the name space and scope, here.

File: pennylane_qrack/qrack_device.py (hunk 2 above)
Comment (WrathfulSpatula) [reply pennylane_qrack/qrack_device.py:213]: Done.

File: setup.py (hunk 1 above)
Comment (WrathfulSpatula) [reply setup.py]: (Outdated.)

**Comment by WrathfulSpatula** [comment]
@cosenal To the point of your comment elsewhere, you asked about examples we could run to show the package works, and I actually ran into an issue with even static linkage of the Qrack library when I attempted to run the `tutorial_qft.py` file in the `PennyLaneAI/qml` repository. I've now spent about a full working day trying to include Qrack as a submodule: it's not easy to do this at all, with `scikit-build` and `setuptools`. Hence, most of the point of this PR is moot, although it's good to review formatting.

Everyone be aware, you _will_ have to install Qrack separately, currently. On the bright side, the Python package installation will complain that the `qrack` library can't be found to link, which _should_ be understood by users with basic computer science savvy. However, let me add explicit instructions to the README! I really wish I could eliminate _every_ point of friction for users, but this is becoming a disproportionate level-of-effort to work around a build process that is theoretically already standardized and working.

**Pull request merged**